### PR TITLE
don't report CPU metrics if it's negative value

### PR DESCRIPTION
**Pull request by nagworld9**
## Description
The cpu usage can be negative value if we notice less current cpu ticks from last time and this could happen when service restart between two polls.

Fix is not to report telemetry if cpu usage is negative value.
 
Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/cgroup.py [hunk 1]
```diff
@@ -241,11 +243,16 @@ def get_throttled_time(self):
...
-        ]
+        tracked = []
+        cpu_usage = self.get_cpu_usage()
+        if cpu_usage >= float(0):
+            tracked.append(MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.PROCESSOR_PERCENT_TIME, self.name, cpu_usage))
+
         if 'track_throttled_time' in kwargs and kwargs['track_throttled_time']:
-            tracked.append(MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.THROTTLED_TIME, self.name, self.get_throttled_time()))
+            throttled_time = self.get_throttled_time()
+            if cpu_usage >= float(0) and throttled_time >= float(0):
```
Comment (nagworld9) [inline azurelinuxagent/common/cgroup.py:253]: In this case, very likely throttled time could have wrong value too. So, I decided not to report both.

File: azurelinuxagent/common/cgroup.py [hunk 2]
```diff
@@ -241,11 +243,16 @@ def get_throttled_time(self):
         return float(self._current_throttled_time - self._previous_throttled_time) / 1E9
 
     def get_tracked_metrics(self, **kwargs):
-        tracked = [
-            MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.PROCESSOR_PERCENT_TIME, self.name, self.get_cpu_usage()),
-        ]
+        tracked = []
+        cpu_usage = self.get_cpu_usage()
```
Comment (nagworld9) [inline azurelinuxagent/common/cgroup.py:247]: updating previous/current values happens inside these get functions on every poll, so no explicit logic needed to reset the values.

**Review by kevinclark19a — APPROVED** [review]

File: azurelinuxagent/common/cgroup.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroup.py:253]: I think we do not need to check for the 2 conditions, if the single condition cpu_usage < 0 is true, we should not report the 2 metrics

File: tests/common/test_cgroupstelemetry.py [hunk 3]
```diff
@@ -374,3 +374,44 @@ def test_extension_telemetry_not_sent_for_empty_perf_metrics(self, *args):  # py
                 metrics = CGroupsTelemetry.poll_all_tracked()
                 self.assertEqual(0, len(metrics))
 
+    @patch("azurelinuxagent.common.cgroup.MemoryCgroup.get_max_memory_usage")
```
Comment (narrieta) [inline tests/common/test_cgroupstelemetry.py:377]: too many mocks here

File: azurelinuxagent/common/cgroup.py (hunk 1 above)
Comment (nagworld9) [reply azurelinuxagent/common/cgroup.py:253]: yeah, if cpu_usage < 0 is true then it won't check second condition. I added second condition for rare case.

File: tests/common/test_cgroupstelemetry.py (hunk 3 above)
Comment (nagworld9) [reply tests/common/test_cgroupstelemetry.py:377]: I think I followed old test cases. Actually, we don't need to mock memory usage as this test/change only for CPU values. I will remove them.

File: tests/common/test_cgroupstelemetry.py [hunk 4]
```diff
@@ -374,3 +374,44 @@ def test_extension_telemetry_not_sent_for_empty_perf_metrics(self, *args):  # py
...
+    @patch("azurelinuxagent.common.cgroup.MemoryCgroup.get_memory_usage")
+    @patch("azurelinuxagent.common.cgroup.CpuCgroup.get_cpu_usage")
+    @patch("azurelinuxagent.common.cgroup.CpuCgroup.get_throttled_time")
+    @patch("azurelinuxagent.common.cgroup.CGroup.is_active")
+    def test_cgroup_telemetry_should_not_report_cpu_negative_value(self, patch_is_active, path_get_throttled_time, patch_get_cpu_usage, patch_get_memory_usage, patch_get_memory_max_usage, *args):# pylint: disable=unused-argument
```
Comment (narrieta) [inline tests/common/test_cgroupstelemetry.py:382]: don't disable unused-argument, use "_" as the argument name instead

File: tests/common/test_cgroupstelemetry.py (hunk 3 above)
Comment (narrieta) [reply tests/common/test_cgroupstelemetry.py:377]: ok, do you need the mock for is_active?

**Review by kevinclark19a — APPROVED** [review]

**Pull request merged**